templates/policygen/org_policies/main.tf

```diff
@@ -17,142 +17,207 @@
 # See the following resources for the details of policies enforced.
 
 terraform {
+  required_version = "~> 0.12.0"
+  required_providers {
+    google      = "~> 3.0"
+    google-beta = "~> 3.0"
+  }
   backend "gcs" {}
 }
 
 # App Engine
-resource "google_organization_policy" "appengine_disable_code_download" {
-  org_id     = var.org_id
-  constraint = "appengine.disableCodeDownload"
-
-  boolean_policy {
-    enforced = true
-  }
+module "orgpolicy_appengine_disable_code_download" {
+  source  = "terraform-google-modules/org-policy/google"
+  version = "~> 3.0.2"
+
+  policy_for      = var.policy_for
+  organization_id = var.policy_for == "organization" ? var.node_id : null
```

Comment:
Instead of repeating these everywhere, why not have a `locals` block up top:
```
locals {
  organization_id = var.policy_for == "organization" ? var.node_id : null
  folder_id       = var.policy_for == "folder" ? var.node_id : null
  project_id      = var.policy_for == "project" ? var.node_id : null
}
...
  organization_id = local.organization_id
  folder_id = local.folder_id
  project_id = local.project_id
```

Reply:
I would actually suggest just using templates here.

{{$policy_for := ""}}
{{$res_field := ""}}
{{$res_id := ""}}
{{if eq .POLICY_FOR "organization"}}
  {{$policy_for = "organization"}}
  {{$policy_for := "organization_id"}}
  {{$res_id = .ORG_ID}}
{{else if eq .POLICY_FOR "folder"}}
  ...
{{else}}
  ...
{{end}}

Then in the module:
module "foo" {
  policy_for: "{{$policy_for}}"
  {{$res_field}} = "{{$res_id}}
}

I am going to use this strategy to use a single component to create foundation components for org, folder and project from a single template.

This way is also superior to direct terraform because you don't need to define unneeded fields so you get more concise configs.

Reply:
Also, do the modules offer any benefit apart from letting you set these resources at different levels? If not, maybe we should just stick to the resources here and use our templates to create the different resource types 

e.g.

resource "google_{{.POLICY_FOR}}_policy" "authorized_networks" {
  {{$res_field}} = "{{$res_id}}"
}

Reply:
One more idea you can also pass these values in from the recipes, so the templates become pretty simple.

Reply:
re: module vs. resource: The module also gives the benefits of easily excluding projects or folders, which is a useful feature to be added later. And the interface is actually more readable imo, so I would prefer module here.

Reply:
For the other topic, I think the config as is + Martin's suggestion for the local vars is actually the cleanest way here. I'm not too sure about using heavy template in `main.tf` here, and the `policy_for` and `node_id` now at least apply across the entire `main.tf`. Maybe later if we are treating each policy individually, then we can put heavy templating in the `main.tf`,

Another thing I just thought about is if we are preparing to give the terraform configs to the users, and they might take it from there, they actually follow the terraform's convention to add stuff in variables.tf and terraform.tfvars. Leaving some common or basic variables in our generated `variables.tf` might be helpful.

Reply:
okay, changed my mind. I'm sold on the template way in `main.tf` for this.

Reply:
variables are fine, i just dont see the benefit of having org/folder/project id checks happening in native terraform or putting every field in every resource when we can use templates to generate cleaner ones

Reply:
I tried out the template way, it's pretty clean.